tools/mirror/src/chain_tracker.rs

```diff
@@ -0,0 +1,246 @@
+        self.num_txs_awaiting_nonce -= 1;
+        t.tx.nonce = nonce;
+        let tx = SignedTransaction::new(
+            t.target_private.sign(&t.tx.get_hash_and_size().0.as_ref()),
+            t.tx,
+        );
+        tracing::debug!(target: "mirror", "prepared a transaction for ({:?}, {:?}) that was previously waiting for the access key to appear on chain",
+        &tx.transaction.signer_id, &tx.transaction.public_key);
+        chunk.txs.push(tx);
+    }
+
+    pub(crate) fn queue_block(&mut self, block: MappedBlock) {
+        self.height_queued = Some(block.source_height);
+        for c in block.chunks.iter() {
+            self.num_txs_awaiting_nonce += c.txs_awaiting_nonce.len();
+        }
+        self.queued_blocks.push_back(block);
+    }
+
+    pub(crate) fn next_batch_time(&self) -> Instant {
+        match &self.send_time {
+            Some(t) => t.as_ref().deadline().into_std(),
+            None => Instant::now(),
+        }
+    }
+
+    pub(crate) async fn next_batch(&mut self) -> Option<MappedBlock> {
+        if let Some(sleep) = &mut self.send_time {
+            sleep.await;
+        }
+        let block = self.queued_blocks.pop_front();
+        if let Some(block) = &block {
+            for c in block.chunks.iter() {
+                self.num_txs_awaiting_nonce -= c.txs_awaiting_nonce.len();
+            }
+        }
+        block
+    }
+
+    fn remove_tx(&mut self, tx: &IndexerTransactionWithOutcome) {
+        let k = (tx.transaction.signer_id.clone(), tx.transaction.public_key.clone());
+        match self.txs_by_signer.entry(k.clone()) {
+            Entry::Occupied(mut e) => {
+                let txs = e.get_mut();
+                if !txs.remove(&TxId { hash: tx.transaction.hash, nonce: tx.transaction.nonce }) {
+                    tracing::warn!(target: "mirror", "tried to remove nonexistent tx {} from txs_by_signer", tx.transaction.hash);
+                }
+                // split off from hash: default() since that's the smallest hash, which will leave us with every tx with nonce
+                // greater than this one in txs_left.
+                let txs_left = txs.split_off(&TxId {
+                    hash: CryptoHash::default(),
+                    nonce: tx.transaction.nonce + 1,
+                });
+                if !txs.is_empty() {
+                    tracing::warn!(target: "mirror", "{} Transactions for {:?} skipped by inclusion of tx with nonce {}: {:?}. These will never make it on chain.",
+                    txs.len(), &k, tx.transaction.nonce, &txs);
+                }
+                *txs = txs_left;
+                if txs.is_empty() {
+                    self.txs_by_signer.remove(&k);
+                }
+            }
+            Entry::Vacant(_) => {
+                tracing::warn!(target: "mirror", "recently removed tx {}, but ({:?}, {:?}) not in txs_by_signer",
+                tx.transaction.hash, tx.transaction.signer_id, tx.transaction.public_key);
```

Comment:
indent
A similar indentation problem is seen in other files in the PR.

Reply:
yeah wonder why cargo fmt doesnt fix that...

Reply:
Because it’s a macro.